internal/action/render.go

```diff
@@ -0,0 +1,235 @@
+		}
+		defer reg.Destroy()
+		r.Registry = reg
+	}
+
+	var cfgs []declcfg.DeclarativeConfig
+	for _, ref := range r.Refs {
+		var (
+			cfg *declcfg.DeclarativeConfig
+			err error
+		)
+		if stat, serr := os.Stat(ref); serr == nil && stat.IsDir() {
+			cfg, err = declcfg.LoadDir(ref)
+		} else {
+			cfg, err = r.imageToDeclcfg(ctx, ref)
+		}
+		if err != nil {
+			return nil, fmt.Errorf("render reference %q: %v", ref, err)
+		}
+		renderBundleObjects(cfg)
+		cfgs = append(cfgs, *cfg)
+	}
+
+	return combineConfigs(cfgs), nil
+}
+
+func (r Render) createRegistry() (*containerdregistry.Registry, error) {
+	cacheDir, err := os.MkdirTemp("", "render-registry-")
+	if err != nil {
+		return nil, fmt.Errorf("create tempdir: %v", err)
+	}
+
+	reg, err := containerdregistry.NewRegistry(
+		containerdregistry.WithCacheDir(cacheDir),
+		containerdregistry.WithLog(nullLogger()),
+	)
+	if err != nil {
+		return nil, err
+	}
+	return reg, nil
+}
+
+func (r Render) imageToDeclcfg(ctx context.Context, imageRef string) (*declcfg.DeclarativeConfig, error) {
+	ref := image.SimpleReference(imageRef)
+	if err := r.Registry.Pull(ctx, ref); err != nil {
+		return nil, err
+	}
+	labels, err := r.Registry.Labels(ctx, ref)
+	if err != nil {
+		return nil, err
+	}
+	tmpDir, err := ioutil.TempDir("", "render-unpack-")
+	if err != nil {
+		return nil, err
+	}
+	defer os.RemoveAll(tmpDir)
+	if err := r.Registry.Unpack(ctx, ref, tmpDir); err != nil {
+		return nil, err
+	}
+
+	var cfg *declcfg.DeclarativeConfig
+	if dbFile, ok := labels[containertools.DbLocationLabel]; ok {
+		cfg, err = sqliteToDeclcfg(ctx, filepath.Join(tmpDir, dbFile))
+		if err != nil {
+			return nil, err
+		}
+	} else if configsDir, ok := labels["operators.operatorframework.io.index.configs.v1"]; ok {
+		// TODO(joelanford): Make a constant for above configs location label
+		cfg, err = declcfg.LoadDir(filepath.Join(tmpDir, configsDir))
+		if err != nil {
+			return nil, err
+		}
+	} else if _, ok := labels[bundle.PackageLabel]; ok {
+		img, err := registry.NewImageInput(ref, tmpDir)
+		if err != nil {
+			return nil, err
+		}
+
+		cfg, err = bundleToDeclcfg(img.Bundle)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		labelKeys := sets.StringKeySet(labels)
+		labelVals := []string{}
+		for _, k := range labelKeys.List() {
+			labelVals = append(labelVals, fmt.Sprintf("  %s=%s", k, labels[k]))
+		}
+		if len(labelVals) > 0 {
+			return nil, fmt.Errorf("render %q: image type could not be determined, found labels\n%s", ref, strings.Join(labelVals, "\n"))
+		} else {
+			return nil, fmt.Errorf("render %q: image type could not be determined: image has no labels", ref)
+		}
+	}
+	return cfg, nil
+}
+
+func sqliteToDeclcfg(ctx context.Context, dbFile string) (*declcfg.DeclarativeConfig, error) {
+	db, err := sqlite.Open(dbFile)
+	if err != nil {
+		return nil, err
+	}
+
+	migrator, err := sqlite.NewSQLLiteMigrator(db)
+	if err != nil {
+		return nil, err
+	}
+	if migrator == nil {
```

Comment:
To be fair, the function you linked could be changed in the future and could result in a function call on a nil pointer if this isn't caught.

Reply:
I copied this code from an existing place in the codebase.

I _generally_ code ([V], error) functions to never return `nil, nil` unless I specifically document that they can. And when I call ([V], error) functions, I make that same assumption. So far, I don't think I've been bitten by nil pointer errors like this.

So :man_shrugging:.. I'll leave it like this in this case because there was already a decision elsewhere to include this check.